File: Code/Chapter 8/VB.NET/Exercise 8.9/Form1.vb
Imports CreateDataSet
Imports System.Data
Imports System.Xml

Public Class Form1
   Private dataFilePath As String = "..\..\..\..\Data.Xml"
   Private WithEvents CustomersTable As DataTable
   Private xdd As XmlDataDocument

   Private Sub Form1_Load( _
     ByVal sender As System.Object, ByVal e As System.EventArgs) _
     Handles MyBase.Load
      CustomersTable = CreateDataSet.DataSetFiller.FillDataset(dataFilePath).Tables("Customers")
      xdd = New XmlDataDocument(CustomersTable.DataSet)
      dgView.DataSource = CustomersTable
      xdd.Save(Application.ExecutablePath & "_xdd.xml")
      xmlViewer.Navigate(Application.ExecutablePath & "_xdd.xml")
   End Sub

   Private Sub CustomersTable_RowChanged( _
     ByVal sender As Object, ByVal e As System.Data.DataRowChangeEventArgs) _
     Handles CustomersTable.RowChanged
      xdd.Save(Application.ExecutablePath & "_xdd.xml")
      xmlViewer.Navigate(Application.ExecutablePath & "_xdd.xml")
   End Sub
End Class

File: Code/Chapter 8/C#/Exercise 8.9/Form1.cs
#region Using directives

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using System.Xml;

#endregion

namespace Exercise_8_9
{
   partial class Form1 : Form
   {
      private static string dataFilePath = @"..\..\..\..\..\Data.Xml";
      private DataTable CustomersTable;
      private XmlDataDocument xdd;

      public Form1()
      {
         InitializeComponent();
      }

      private void Form1_Load(object sender, EventArgs e)
      {
         CustomersTable = CreateDataSet.DataSetFiller.FillDataset(dataFilePath).Tables["Customers"];
         xdd = new XmlDataDocument(CustomersTable.DataSet);
         dgView.DataSource = CustomersTable;
         xdd.Save(Application.ExecutablePath + "_xdd.xml");
         xmlViewer.Navigate(Application.ExecutablePath + "_xdd.xml");
         CustomersTable.RowChanged +=new DataRowChangeEventHandler(CustomersTable_RowChanged);
      }

      void CustomersTable_RowChanged(object sender, DataRowChangeEventArgs e)
      {
         xdd.Save(Application.ExecutablePath + "_xdd.xml");
         xmlViewer.Navigate(Application.ExecutablePath + "_xdd.xml");
      }
   }
}